---
title: "Coloc_analysis"
author: "Emilia Vartiainen"
date: "6/23/2020"
output: pdf_document
---

```{r libraries, include=FALSE}
library("dplyr")
library("ggplot2")
library("readr")
library("coloc")
library("GenomicRanges")
library("Rsamtools")
```

# Load eQTL catalogues 

``` {r load_data}
tabix_paths = read.delim("https://raw.githubusercontent.com/eQTL-Catalogue/eQTL-Catalogue-resources/master/tabix/tabix_ftp_paths.tsv", 
                         sep = "\t", header = TRUE, stringsAsFactors = FALSE) %>% dplyr::as_tibble()
imported_tabix_paths = read.delim("https://raw.githubusercontent.com/eQTL-Catalogue/eQTL-Catalogue-resources/master/tabix/tabix_ftp_paths_imported.tsv", 
                                  sep = "\t", header = TRUE, stringsAsFactors = FALSE) %>% dplyr::as_tibble()
```

# A general function to quickly import tabix indexed tab-separated files into data_frame

``` {r scanTabixDataFrame}
#' @param tabix_file Path to tabix-indexed text file
#' @param param An instance of GRanges, RangedData, or RangesList
#' provide the sequence names and regions to be parsed. Passed onto Rsamtools::scanTabix()
#' @param ... Additional parameters to be passed on to readr::read_delim()
#' @return List of data_frames, one for each entry in the param GRanges object.

scanTabixDataFrame <- function(tabix_file, param, ...){
  tabix_list = Rsamtools::scanTabix(tabix_file, param = param)
  df_list = lapply(tabix_list, function(x,...){
    if(length(x) > 0){
      if(length(x) == 1){
        #Hack to make sure that it also works for data frames with only one row
        #Adds an empty row and then removes it
        result = paste(paste(x, collapse = "\n"),"\n",sep = "")
        result = readr::read_delim(result, delim = "\t", ...)[1,]
      }else{
        result = paste(x, collapse = "\n")
        result = readr::read_delim(result, delim = "\t", ...)
      }
    } else{
      #Return NULL if the nothing is returned from tabix file
      result = NULL
    }
    return(result)
  }, ...)
  return(df_list)
}
```

# In eQTL Catalogue, variants with multiple rsids are split over multiple rows in the summary statistics files. Thus, we first want to retain only one unique record per variant. To simplify colocalisation analysis, we also want to exclude multi-allelic variants. The following function imports summary statistics from a tabix-index TSV file and performs necessary filtering.

``` {r import_eQTLcatalogue}
import_eQTLCatalogue <- function(ftp_path, region, column_names){
  #Fetch summary statistics with Rsamtools
  summary_stats = scanTabixDataFrame(ftp_path, region, col_names = column_names)[[1]] 
  
  # Should I remove the duplicates?
  # Remove rsid duplicates and multi-allelic variant
  #summary_stats = dplyr::select(summary_stats, -rsid) %>% 
  #  dplyr::distinct() %>% #rsid duplicates
  #  dplyr::mutate(id = paste(chromosome, position, sep = ":")) %>% 
  #  dplyr::group_by(id) %>% 
  #  dplyr::mutate(row_count = n()) %>% dplyr::ungroup() %>% 
  #  dplyr::filter(row_count == 1) #Multialllics
  
  return(summary_stats)
}
```

# Determine the genomic loci we're using in colc analysis

``` {r region_ranges}
top_variants <- read.csv("/Users/eahvarti/Documents/coloc_analysis/Data/I9_VARICVE.top.out", sep = "\t", header = TRUE)
top_variants <- top_variants[which(top_variants$chr != 23),]

region_granges = GenomicRanges::GRanges(
  seqnames = as.vector(top_variants$chr), 
  ranges = IRanges::IRanges(start = as.vector(top_variants$start), end = as.vector(top_variants$end)), 
  strand = "*") 
```

# Determine the eQTL dataset we're using and select the genonic loci, they are in summary_stats variable 

``` {r eQTL_summary_stats}
eQTL_data_name = dplyr::filter(tabix_paths, study == "CEDAR", tissue_label == "platelet") # Specify which eQTL dataset is going to be used 

#Extract column names from first file
column_names = colnames(readr::read_tsv(eQTL_data_name$ftp_path[1])) 

summary_stats = import_eQTLCatalogue(eQTL_data_name$ftp_path, region_granges, column_names) # These are the eQTL summary statistics for "region_granges" genomic loci
```



















