---
title: "Coloc Wrapper Example"
author: "Emilia Vartiainen"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Goal 

- Non coding GWAS hit, we want to know if it appears in the eQTL data
- The region of interest is 5:128070567-128070567, gene SLC12A2.

## Download the data
### 1. GWAS from FinnGen
Varicose veins, summary statistics
```
gsutil -m cp gs://finngen-production-library-green/finngen_R6/finngen_R6_analysis_data/summary_stats/release/finngen_R6_I9_VARICVE.gz .

gsutil -m cp gs://finngen-production-library-green/finngen_R6/finngen_R6_analysis_data/summary_stats/release/finngen_R6_I9_VARICVE.gz.tbi .
```

### 2. eQTL from the eQTL catalogue 
BLUEPRINT study, monocytes tissue, summary statistics
```
wget ftp://ftp.ebi.ac.uk/pub/databases/spot/eQTL/csv/BLUEPRINT/exon/BLUEPRINT_exon_monocyte.all.tsv.gz
wget ftp://ftp.ebi.ac.uk/pub/databases/spot/eQTL/csv/BLUEPRINT/exon/BLUEPRINT_exon_monocyte.all.tsv.gz.tbi
```

## Specify the parameters
- Region of interest is `5:128070567-128070567`. We're adding a 1500000 bp window. 
- Sample size for the eQTL data set can be found [here](https://www.ebi.ac.uk/eqtl/Studies/)
- Number of cases and controls for the GWAS data can be found
```
gsutil cat gs://finngen-production-library-green/finngen_R6/finngen_R6_analysis_data/finngen_R6_pheno_n.tsv | grep "I9_VARICVE"
```

## Trimming the data
Trimming GWAS and eQTL data according to the region of interest.

- `file`: GWAS or eQTL file path or url
- `region`: genomic region of interest, format `chr:start-end`
- `out`: output file 

### 1. GWAS
```
Rscript extdata/step1_subset_data.R	\
	--file=finngen_R6_I9_VARICVE.gz\
	--region="5:126570567-129570567" \
  --out=finngen_R6_I9_VARICVE_5_128070567_128070567.txt/
```

#### Add ids to GWAS data
Ids need to be in format `chr#_chrom_pos_alt_ref`
```
cat finngen_R6_I9_VARICVE_5_128070567_128070567.txt | awk '{print "chr"$1"_"$2"_"$3"_"$4"\t"$0}'  > tmp && mv tmp finngen_R6_I9_VARICVE_5_128070567_128070567.txt
```

### 2. eQTL
```
Rscript extdata/step1_subset_data.R	\
	--file=BLUEPRINT_exon_monocyte.all.tsv.gz\
	--region="5:126570567-129570567" \
  --out=BLUEPRINT_exon_monocyte_5_128070567_128070567.txt/
```

## Running coloc

- `eqtl`: eQTL summary statistics file for one region
- `gwas`: GWAS summary statistics file for one region
- `header_eqtl`: header of eQTL file, named vector in quotes
- `header_gwas`: header of GWAS file, named vector in quotes 
- `info_eqtl`: options for eQTL dataset, more info [here](https://www.rdocumentation.org/packages/coloc/versions/3.2-1/topics/coloc.abf)
  - `type`: the type of data in dataset - either "quant" or "cc" to denote quantitative or case-control
  - `sdY`: for a quantitative trait, the population standard deviation of the trait
  - `N`: number of samples
- `info_gwas`: options for GWAS dataset, more info [here](https://www.rdocumentation.org/packages/coloc/versions/3.2-1/topics/coloc.abf)
  - `type`: the type of data in dataset - either "quant" or "cc" to denote quantitative or case-control
  - `s`: for a case control dataset, the proportion of samples in dataset that are cases
  - `N`: number of samples
- `p1`: the prior probability that any random SNP in the region is associated with exactly trait 1
- `p2`: the prior probability that any random SNP in the region is associated with exactly trait 2
- `p3`: the prior probability that any random SNP in the region is associated with both traits
- `locuscompare_thresh`: PP4 threshold that plots the locuscompare plot 
- `out`: output file

```
Rscript extdata/step2_run_coloc.R	\
	--eqtl="BLUEPRINT_exon_monocyte_5_128070567_128070567.txt" \
	--gwas="finngen_R6_I9_VARICVE_5_128070567_128070567.txt" \
	--header_eqtl="c(varid = 'variant', pvalues = 'pvalue', MAF = 'maf', gene_id = 'molecular_trait_id')" \
	--header_gwas="c(varid = 'chr#chrom_pos_ref_alt', pvalues = 'pval', MAF = 'af_alt')" \
	--info_gwas="list(type = 'cc', s = 20425/(20425+225735), N  = 20425+225735)" \
	--info_eqtl="list(type = 'quant', sdY = 1, N = 554)" \
	--p1=1e-4 \
	--p2=1e-4 \
	--p12=1e-5 \
	--locuscompare_thresh=0.7 \
	--out="coloc_VARICVE_BLUEPRINT_example_5_128070567_128070567.txt" \
```

## Evaluating the results
### Coloc results
Columns:

- `gene`: gene ID
- `nsnps`: number of SNPs analysed
- `PP.H0.abf`: posterior probabilitiy of H0 (no causal variant)
- `PP.H1.abf`: posterior probability of H1 (causal variant for trait 1 only)
- `PP.H2.abf`: posterior probability of H2 (causal variant for trait 2 only)
- `PP.H3.abf`: posterior probability of H3 (two distinct causal variants)
- `PP.H4.abf`: posterior probability of H4 (one common causal variant)

```{r, echo=FALSE}
library(knitr)
coloc_results <- data.table::fread("/Users/eahvarti/Documents/R_studio/Coloc_analysis/TESTING/coloc_VARICVE_BLUEPRINT_example_5_128070567_128070567.txt", sep = "\t")
table <- coloc_results[118:125,]
knitr::kable(table)
```

### Locuscompare plot

```{r, echo=FALSE, out.width = "80%", fig.align = "center"}
knitr::include_graphics("/Users/eahvarti/Documents/R_studio/Coloc_analysis/TESTING/coloc_VARICVE_BLUEPRINT_example_5_128070567_128070567_locuscompare_ENSG00000064651.14_5_128114212_128114287.png")
```










