---
title: "Coloc-wrapper usage"
output: 
  html_document:
  latex_engine: xelatex
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### Introduction
Colocalization analysis is used to detect genetic causality between two different GWAS traits. Coloc-wrapper performs genetic colocalization analysis for GWAS and eQTL datasets in a given region using `coloc.abf()` function from Coloc R-package. It calculates posterior probabilities for the following five hypothesis for each gene in the region under the assumption of a single causal variant for each trait:

$H_0$: no association <br />
$H_1$: association to trait 1 only <br />
$H_2$: association to trait 2 only <br />
$H_3$: association to both traits, distinct causal variants <br />
$H_4$: association to both traits, shared causal variant <br />

The posterior probability of hypothesis 4, PP4, determines the possible colocalization. A common threshold for it is PP4 > 0.8.

### Getting started
#### I on local machine
- R version >3.6.2
- R-packages: "coloc", "data.table", "ggplot2", "optparse", "R.utils"
- tabix

#### II using docker
Sina

### Data 
The input files are the following:

#### 1. GWAS summary statistics


#### 2. eQTL summary statistics
- eQTL data can be found here: https://www.ebi.ac.uk/eqtl/ 

#### 3. Region

### Usage
Running coloc-wrapper involves two steps: <br />
1. Trimming data, both GWAS and eQTL, according to a predefined region <br />
2. Running coloc <br />

#### 1. Trimming data
- `file`: GWAS or eQTL file path or url
- `region`: genomic region of interest, format `chr:start-end`
- `out`: output file 

```
Rscript extdata/step1_subset_data.R	\
	--file=ftp://ftp.ebi.ac.uk/pub/databases/spot/eQTL/csv/Lepik_2017/ge/Lepik_2017_ge_blood.all.tsv.gz \
	--region="1:10565520-10965520" \
  --out=tmp.txt
```

#### 2. Run coloc 
- `eqtl`: eQTL summary statistics file for one region
- `gwas`: GWAS summary statistics file for one region
- `header_eqtl`: header of eQTL file, named vector in quotes
- `header_gwas`: header of GWAS file, named vector in quotes 
- `info_eqtl`: options for eQTL dataset:
  - `type`: the type of data in dataset - either "quant" or "cc" to denote quantitative or case-control
  - `s`: for a case control dataset, the proportion of samples in dataset that are cases
  - 
- `info_gwas`: options for GWAS dataset 
  - ASK ABOUT THESE HOW MUCH TO WRITE
- `p1`: the prior probability that any random SNP in the region is associated with exactly trait 1
- `p2`: the prior probability that any random SNP in the region is associated with exactly trait 2
- `p3`: the prior probability that any random SNP in the region is associated with both traits
- `locuscompare_thresh`: PP4 threshold that plots the locuscompare plota 
- `out`: output file

```
Rscript extdata/step2_run_coloc.R	\
	--eqtl="extdata/Lepik_2017_ge_blood_chr1_ENSG00000142655_ENSG00000130940.all.tsv" \
	--gwas="extdata/I9_VARICVE_chr1.tsv.gz" \
	--header_eqtl="c(varid = 'rsid', pvalues = 'pvalue', MAF = 'maf', gene_id = 'gene_id')" \
	--header_gwas="c(varid = 'rsids', pvalues = 'pval', MAF = 'maf')" \
	--info_gwas="list(type = 'cc', s = 11006/117692, N  = 11006 + 117692)" \
	--info_eqtl="list(type = 'quant', sdY = 1, N = 491)" \
	--p1=1e-4 \
	--p2=1e-4 \
	--p12=1e-5 \
	--locuscompare_thresh=0.8 \
	--out="Coloc_example.txt" \
```
### Output 





# Notes

links:
https://github.com/weizhouUMICH/SAIGE
https://github.com/n-mounier/bGWAS
https://cnsgenomics.com/software/gcta/#COJO
https://chr1swallace.github.io/coloc/
https://devguide.ropensci.org/building.html#readme

## Introduction
What does the method do

## Getting started
### 1 on local machine
how to set up machine to run coloc
which library, R version
### 2 using docker
(tbd sina)

## Data
where can people find eqtl data
minimal requirments

## Minimal example
explain input
GOal: have for a particular region all colocalizations for  a particular eqtl datset an all genes:
Run Rscript extdata/step1_subset_data.R for the GWAS dataset and once for the eqtl dataset. 
Rscript extdata/step2_run_coloc.R
if other datasets or regions are needed: loop. 
explain output

## More complicated example
with changing parameters or locuscompare plot. 










